src/Routing/UrlGeneratorFragmentWrapper.php

```diff
@@ -33,16 +33,17 @@ public function __construct(UrlGeneratorInterface $wrapped)
      */
     public function generate($name, $parameters = [], $referenceType = UrlGeneratorInterface::ABSOLUTE_PATH)
     {
-        $fragment = null;
-        if (isset($parameters['#'])) {
+        $fragment = isset($parameters['_fragment']) ? $parameters['_fragment'] : null;
+        unset($parameters['_fragment']);
+        if ($fragment === null && isset($parameters['#'])) {
```

Comment:
Moved the `unset()` down — as it's harmless regardless — to avoid the need for more condition(al)s